### PR TITLE
bpo-46481: Implement vectorcall for weakref.ref.__call__ method.

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue46481

### CODE REVIEW

**Comment by corona10** [comment]
@vstinner 

>  buildbot/AMD64 Windows8.1 Refleaks PR — Build done.

This failure is not related to this PR :)

File: Objects/weakrefobject.c [hunk 1]
```diff
@@ -128,19 +130,22 @@ gc_clear(PyWeakReference *self)
...
+        return NULL;
+    }
+    Py_ssize_t nargs = PyVectorcall_NARGS(nargsf);
+    if (!_PyArg_CheckPositional("__call__", nargs, 0, 0)) {
+        return NULL;
     }
-    return NULL;
+    PyObject *obj = PyWeakref_GET_OBJECT(self);
+    Py_INCREF(obj);
+    return obj;
```
Comment (kumaraditya303) [inline Objects/weakrefobject.c]:
```suggestion
    return Py_NewRef(obj);
```

File: Objects/weakrefobject.c [hunk 2]
```diff
@@ -374,42 +379,44 @@ _PyWeakref_RefType = {
...
-    0,                          /*tp_base*/
-    0,                          /*tp_dict*/
-    0,                          /*tp_descr_get*/
-    0,                          /*tp_descr_set*/
-    0,                          /*tp_dictoffset*/
-    weakref___init__,           /*tp_init*/
-    PyType_GenericAlloc,        /*tp_alloc*/
-    weakref___new__,            /*tp_new*/
-    PyObject_GC_Del,            /*tp_free*/
+    weakref_dealloc,                        /*tp_dealloc*/
```
Comment (vstinner) [inline Objects/weakrefobject.c]: Can you please leave other lines unchanged to make the commit smaller?

Or if you can motivated, rewrite the type to use the ".tp_member = value" syntax (and emit members which use the default 0 value).

File: Objects/weakrefobject.c [hunk 3]
```diff
@@ -374,42 +379,44 @@ _PyWeakref_RefType = {
...
+    offsetof(PyWeakReference, vectorcall),  /*tp_vectorcall_offset*/
```
Comment (vstinner) [inline Objects/weakrefobject.c]: Is it possible to use .tp_vectorcall instead, to not make _PyWeakReference larger?

Comment (corona10) [reply Objects/weakrefobject.c]: if Py_TPFLAGS_HAVE_VECTORCALL is set, tp_vectorcall_offset should be set.

File: Objects/weakrefobject.c [hunk 4]
```diff
@@ -371,45 +374,25 @@ static PyMethodDef weakref_methods[] = {
...
-    0,                          /*tp_descr_get*/
-    0,                          /*tp_descr_set*/
-    0,                          /*tp_dictoffset*/
-    weakref___init__,           /*tp_init*/
-    PyType_GenericAlloc,        /*tp_alloc*/
-    weakref___new__,            /*tp_new*/
-    PyObject_GC_Del,            /*tp_free*/
+    .tp_name = "weakref",
+    .tp_basicsize = sizeof(PyWeakReference),
+    .tp_itemsize = 0,
```
Comment (vstinner) [inline Objects/weakrefobject.c]: You can omit this one, no?

File: Objects/weakrefobject.c [hunk 5]
```diff
@@ -128,19 +130,20 @@ gc_clear(PyWeakReference *self)
 
 
 static PyObject *
-weakref_call(PyWeakReference *self, PyObject *args, PyObject *kw)
+weakref_vectorcall(PyObject *callable, PyObject *const *args,
```
Comment (vstinner) [inline Objects/weakrefobject.c]: At the ABI level, ``PyWeakReference*`` and ``PyObject*`` are just pointers, so you can use directly ``PyWeakReference*``, no? I don't think that clang Control Flow Integrity (CFI) cares about that, does it?

File: Objects/weakrefobject.c (hunk 4 above)
Comment (corona10) [reply Objects/weakrefobject.c]: yes we can

File: Objects/weakrefobject.c [hunk 6]
```diff
@@ -29,6 +30,7 @@ init_weakref(PyWeakReference *self, PyObject *ob, PyObject *callback)
     self->wr_next = NULL;
     Py_XINCREF(callback);
     self->wr_callback = callback;
```
Comment (vstinner) [inline Objects/weakrefobject.c]: While we are here...

```suggestion   
    self->wr_callback = Py_XNewRef(callback);
```

File: Objects/weakrefobject.c [hunk 7]
```diff
@@ -371,45 +373,24 @@ static PyMethodDef weakref_methods[] = {
...
-    PyType_GenericAlloc,        /*tp_alloc*/
-    weakref___new__,            /*tp_new*/
-    PyObject_GC_Del,            /*tp_free*/
+    .tp_name = "weakref",
+    .tp_basicsize = sizeof(PyWeakReference),
+    .tp_dealloc = weakref_dealloc,
+    .tp_vectorcall_offset = offsetof(PyWeakReference, vectorcall),
+    .tp_repr = (reprfunc)weakref_repr,
+    .tp_hash = (hashfunc)weakref_hash,
+    .tp_call = PyVectorcall_Call,
```
Comment (vstinner) [inline Objects/weakrefobject.c]: Can you please move it closer to tp_vectorcall_offset?

**Review by vstinner — APPROVED** [review]
LGTM. I just added minor coding style remarks. The overall change LGTM ;-) It's a nice optimization!

File: Objects/weakrefobject.c [hunk 8]
```diff
@@ -29,6 +30,7 @@ init_weakref(PyWeakReference *self, PyObject *ob, PyObject *callback)
     self->wr_next = NULL;
     Py_XINCREF(callback);
     self->wr_callback = callback;
+    self->vectorcall = weakref_vectorcall;
```
Comment (vstinner) [inline Objects/weakrefobject.c]: To fix compiler warnings:

```suggestion
    self->vectorcall = (vectorcallfunc)weakref_vectorcall;
```

File: Objects/weakrefobject.c [hunk 9]
```diff
@@ -128,19 +130,19 @@ gc_clear(PyWeakReference *self)
...
+weakref_vectorcall(PyWeakReference *self, PyObject *const *args,
+                   size_t nargsf, PyObject *kwnames)
 {
-    static char *kwlist[] = {NULL};
-
-    if (PyArg_ParseTupleAndKeywords(args, kw, ":__call__", kwlist)) {
-        PyObject *object = PyWeakref_GET_OBJECT(self);
-        Py_INCREF(object);
-        return (object);
+    if (!_PyArg_NoKwnames("__call__", kwnames)) {
```
Comment (vstinner) [inline Objects/weakrefobject.c]: Mmmm. the first parameter should be the function name. Should it be "weakref"? Same remark for _PyArg_CheckPositional().

**Review by vstinner — APPROVED** [review]
LGTM. Now it's perfect! :-D

**Pull request merged**